### ISSUE TITLE
Allow multithreading during mpz operations?

### ISSUE THREAD

**Issue by jepler** [aleaxit/gmpy#291]
Isn't it safe to allow threads during long-running mpz operations?  For instance
```c
diff --git a/src/gmpy2_mul.c b/src/gmpy2_mul.c
index e312769..e147508 100644
--- a/src/gmpy2_mul.c
+++ b/src/gmpy2_mul.c
@@ -45,7 +45,9 @@ GMPy_Integer_MulWithType(PyObject *x, int xtype, PyObject *y, int ytype,
 
     if (IS_TYPE_MPZANY(xtype)) {
         if (IS_TYPE_MPZANY(ytype)) {
+            Py_BEGIN_ALLOW_THREADS;
             mpz_mul(result->z, MPZ(x), MPZ(y));
+            Py_END_ALLOW_THREADS;
             return (PyObject*)result;
         }
```

This is of interest in a program of mine, as individual operations are slow but I have a number of them to do in parallel.  Making this change locally (and to add) lets me get about a 7x speed-up on my particular workload.

Am I overlooking something about how this would interact with the Python GIL? If not, is there any interest in a PR along these lines?

**Comment by casevh** [comment]
I looked at this a few years ago. IIRC, the test results were unpredictable with threading but predictable with multiple process. My test was easily done on multiple processes so it really isn't fair. And my changes to the threading model have been make. I think it is worth trying again. 

I'd like to avoid slowing down the fast operations. A couple of ideas:

1. Use a context flag to enable threading support (release_gil ??) with a default of False. 
2. To avoid releasing the GIL for small operands, we may want to add a check for the length of the underlying mpz_t.

I like the option of a context flag that defaults to False for the soon-to-be-released 2.1.0 version.

A PR would be great.

**Comment by casevh** [comment]
Hi,

I will add this to the next release. I need to make a final release of the current code and then consider it a bug-fix only version. I don't want to add a change of this scope this late in the process. 

Case

**Comment by jepler** [comment]
That's excellent, thank you! Immediately after filing this issue I second-guessed whether it could be as simple as it seemed.

I have found that for my particular "application" (computing large numbers for their own sake, nothing practical) that there is a range of numbers where, particularly for multiplication, it helps to perform multiplication in a thread with ThreadPool.imap. Below the threshold the individual operations are too fast, and above the threshold the numbers are big enough that they don't fit in RAM and swapping begins. With the process model, transferring the representation of the number appears to have an impact on performance.

But thanks to gmpy I have been able to calculate the 2**37th Fibonacci number :stuck_out_tongue_winking_eye: It's about 28.7 billion digits long.
